### PR TITLE
Implementation for commit notification events

### PR DESCRIPTION
**Pull request by manish-sethi**
Signed-off-by: manish <manish.sethi@gmail.com>

#### Type of change
- New feature

#### Description
This PR implements commit notifications for embedded gateway.

### CODE REVIEW

**Comment by sykesm** [comment]
@manish-sethi, please take a look at the test failure to see if it's related. It's not a common flake.

```
--- FAIL: Test_preprocessProtoBlock_processNonEndorserTx (0.00s)
    batch_preparer_test.go:671: 
        	Error Trace:	batch_preparer_test.go:671
        	Error:      	Not equal: 
        	            	expected: []*validation.TxStatInfo{(*validation.TxStatInfo)(0xc000c0e060)}
        	            	actual  : []*validation.TxStatInfo{(*validation.TxStatInfo)(0xc0003d3e60)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,3 +2,3 @@
        	            	  (*validation.TxStatInfo)({
        	            	-  TxIDFromChannelHeader: (string) "",
        	            	+  TxIDFromChannelHeader: (string) (len=3) "tx1",
        	            	   ValidationCode: (peer.TxValidationCode) 0,
        	Test:       	Test_preprocessProtoBlock_processNonEndorserTx
```

File: core/ledger/kvledger/kv_ledger.go [hunk 1]
```diff
@@ -887,14 +893,64 @@ func (l *kvLedger) GetMissingPvtDataTracker() (ledger.MissingPvtDataTracker, err
...
-	return nil, errors.New("yet-to-implement")
+	l.commitNotifierLock.Lock()
+	defer l.commitNotifierLock.Unlock()
+
+	if l.commitNotifier != nil {
+		return nil, errors.New("only one commit notifications channel is allowed at a time")
+	}
+
+	l.commitNotifier = &commitNotifier{
+		dataChannel: make(chan *ledger.CommitNotification, 10),
```
Comment (sykesm) [inline core/ledger/kvledger/kv_ledger.go:916]: Can you explain why you chose `10` for the buffer size? I don't have an issue with it but I'm wondering where it came from.

File: core/ledger/kvledger/kv_ledger.go [hunk 2]
```diff
@@ -64,6 +64,9 @@ type kvLedger struct {
 	// reconciliation and may be updated during a regular block commit.
 	// Hence, we use atomic value to ensure consistent read.
 	isPvtstoreAheadOfBlkstore atomic.Value
+
+	commitNotifierLock sync.RWMutex
```
Comment (sykesm) [inline core/ledger/kvledger/kv_ledger.go]: You're using an `RWMutex` but I don't see any calls to `RLock`. Should this be a simple `sync.Mutex`?

File: core/ledger/kvledger/kv_ledger.go [hunk 3]
```diff
@@ -887,14 +893,64 @@ func (l *kvLedger) GetMissingPvtDataTracker() (ledger.MissingPvtDataTracker, err
...
+		close(l.commitNotifier.dataChannel)
+		l.commitNotifier = nil
+	default:
+		txIDValidationCodes := map[string]peer.TxValidationCode{}
+
+		for _, t := range txStatsInfo {
+			txID := t.TxIDFromChannelHeader
+			_, ok := txIDValidationCodes[txID]
+
+			if txID == "" || ok {
```
Comment (sykesm) [inline core/ledger/kvledger/kv_ledger.go:942]: Under what conditions would we see an empty `txID`?

File: core/ledger/kvledger/kv_ledger_test.go [hunk 4]
```diff
@@ -1118,6 +1119,144 @@ func TestCollectionConfigHistoryRetriever(t *testing.T) {
...
+}
+
+func TestCommitNotificationsOnBlockCommit(t *testing.T) {
+	conf, cleanup := testConfig(t)
+	defer cleanup()
+	provider := testutilNewProvider(conf, t, &mock.DeployedChaincodeInfoProvider{})
+	defer provider.Close()
+
+	bg, gb := testutil.NewBlockGenerator(t, "testLedger", false)
+	protoutil.BlockHeaderHash(gb.Header)
```
Comment (sykesm) [inline core/ledger/kvledger/kv_ledger_test.go]: What's this for? It's calculating the hash on the block bytes but throwing it away.

**Review by sykesm — COMMENTED** [review]
LGTM outside of a couple of questions and the test failure.

**Comment by manish-sethi** [comment]
> @manish-sethi, please take a look at the test failure to see if it's related. It's not a common flake.

It was from this commit itself. I missed including this change before  submitting the PR.

File: core/ledger/kvledger/kv_ledger.go (hunk 1 above)
Comment (manish-sethi) [reply core/ledger/kvledger/kv_ledger.go:916]: No special calculation. Just to give enough buffer if at times gateway becomes slow in consuming the events, the commits are not blocked.

File: core/ledger/kvledger/kv_ledger.go (hunk 2 above)
Comment (manish-sethi) [reply core/ledger/kvledger/kv_ledger.go]: Done

File: core/ledger/kvledger/kv_ledger.go (hunk 3 above)
Comment (manish-sethi) [reply core/ledger/kvledger/kv_ledger.go:942]: As far as I know, the orderer does not perform a check on txid.

File: core/ledger/kvledger/kv_ledger_test.go (hunk 4 above)
Comment (manish-sethi) [reply core/ledger/kvledger/kv_ledger_test.go]: done

**Review by sykesm — APPROVED** [review]

**Pull request merged**